File: binary_search.rb
# frozen_string_literal: true

# Binary Search for int & strings in array
class BinarySearch
  attr_accessor :arr, :element
  attr_reader :moves

  def initialize(arr:, element:)
    @arr = arr
    @element = element
  end

  def integer_search
    high = arr.size - 1
    low = 0
    moves = 0
    while low <= high
      mid = (low + high) / 2
      guess = arr[mid]
      moves += 1
      guess > element ? (high = mid - 1) : (low = mid + 1)

      return "Your number is #{element} with #{moves} moves" if guess == element
    end
  end
end

binario = BinarySearch.new(arr: (1..10).to_a, element: 5)
puts binario.integer_search

File: binary_search.js
'use strict';

function BinarySearch(list, item) {
  const low = 0;
  let high = list.length - 1;

  while (low <= high) {
    const mid = (low + high);
    const guess = list[mid];
    if (guess === item) {
      return mid;
    } else if (guess > item) {
      high = mid - 1;
    } else {
      high = mid + 1;
    }
  }
  return "None";
}

const my_list = [1, 3, 5, 7, 9]
console.log(BinarySearch(my_list, 7))
console.log(BinarySearch(my_list, -1))